tpie/packed_array.h

```diff
@@ -181,6 +181,12 @@ class packed_array: public linear_memory_base<packed_array<T, B> >{
 		iter_return_type * operator->() {return &elm;}
 		iter_base & operator=(const iter_base & o) {elm.index = o.elm.index; elm.elms=o.elm.elms; return *this;}
 		iter_base(iter_base const &o): elm(o.elm) {};
+
+		iter_return_type & operator*() const
+		{ return const_cast<iter_return_type&>(elm); }
+
+		iter_return_type * operator->() const
+		{ return const_cast<iter_return_type*>(&elm); }
```

Comment:
This const_cast seems a bit suspicious to me. The following would be better - does it not work?

```c++
		const iter_return_type & operator*() const
		{ return elm; }

		const iter_return_type * operator->() const
		{ return &elm; }
```

Reply:
Hmm, now that you point it out, I'm quite confused about Steffan from 7 months ago too. I'll do that change (which I agree with too) and we'll see whether all tests pass (as much as they did earlier).

Reply:
Interestingly, we then get from MSVC 14.34.31933
```
error C2678: binary '=': no operator found which takes a left-hand operand of type 'const tpie::packed_array<bool,1>::iter_return_type' (or there is no acceptable conversion)
```
Here is the MSVC sorting algorithm that is the culprit
```cpp
template <class _BidIt, class _Pr>
_CONSTEXPR20 _BidIt _Insertion_sort_unchecked(const _BidIt _First, const _BidIt _Last, _Pr _Pred) {
    // insertion sort [_First, _Last)
    if (_First != _Last) {
        for (_BidIt _Mid = _First; ++_Mid != _Last;) { // order next element
            _BidIt _Hole               = _Mid;
            _Iter_value_t<_BidIt> _Val = _STD move(*_Mid);

            if (_DEBUG_LT_PRED(_Pred, _Val, *_First)) { // found new earliest element, move to front
                _Move_backward_unchecked(_First, _Mid, ++_Hole);
                *_First = _STD move(_Val);
            } else { // look for insertion point after first
                for (_BidIt _Prev = _Hole; _DEBUG_LT_PRED(_Pred, _Val, *--_Prev); _Hole = _Prev) {
                    *_Hole = _STD move(*_Prev); // move hole down
                }

                *_Hole = _STD move(_Val); // insert element in hole
            }
        }
    }

    return _Last;
}
```
Notice the lines with `*_First = _STD move(_Val);`, and `*_Hole = ...`. Alternatively, the issue might be that the packed array is `const` in the first place.

Reply:
Very interesting.

* Does this compile error arise in an open source project when using MSVC and tpie? In other words, can you point me to the code where the problem occurs?
* Does MSVC provide a template instantiation traceback, which could maybe show us how the packed array iterator becomes `const`? It seems like a range of const iterators indeed shouldn't be sortable, but maybe the const-ness is induced wrongly higher up in the call stack.

Reply:
- This is from compiling the TPIE unit tests (see item *v.* in the trace below)
- Here is the log from the compiler output as it is available through the GitHub Action log. I took the liberty to clean it up/format it, so we actually can understand what is going on.
  1. *include\algorithm(7517)*: error C2678: binary '=': no operator found which takes a left-hand operand of type `const tpie::packed_array<bool,1>::iter_return_type`

   2. *include\algorithm(7639)*: while trying to match the argument list `(const tpie::packed_array<bool,1>::iter_return_type, tpie::packed_array<bool,1>::iter_return_type)` for the function template instantiation `_BidIt std::_Insertion_sort_unchecked<_RanIt,_Pr>(const _BidIt, const _BidIt,_Pr)` being compiled with
        ```
        [
            _BidIt=tpie::packed_array<bool,1>::iter_base<true>,
            _RanIt=tpie::packed_array<bool,1>::iter_base<true>,
            _Pr=std::less<void>
        ]
        ```
    3. *include\algorithm(7669)*: see reference to function template instantiation `void std::_Sort_unchecked<_RanIt,_Fn>(_RanIt,_RanIt,__int64,_Pr)` being compiled with
       ``` 
       [
            _RanIt=tpie::packed_array<bool,1>::iter_base<true>,
            _Fn=std::less<void>,
            _Pr=std::less<void>
       ]
       ```
    4. *include\algorithm(7674)*: see reference to function template instantiation `void std::sort<_RanIt,std::less<void>>(const _RanIt, const _RanIt,_Pr)` being compiled with
        ```
        [
            _RanIt=tpie::packed_array<bool,1>::iter_base<true>,
            _Pr=std::less<void>
        ]
        ```
     5. *tpie\test\unit\test_array.cpp(223)*: see reference to function template instantiation `void std::sort<tpie::packed_array<bool,1>::iter_base<true>>(const _RanIt, const _RanIt)` being compiled with
        ```
        [
            _RanIt=tpie::packed_array<bool,1>::iter_base<true>
        ]
        ```
  It seems like the `const` is added directly with the `std::sort` call in the unit test. The `bit_array` `hat` in the unit test in question is *not* set to const.

  Either this is because the MSVC 14.34 STL provides a [const sort](https://github.com/microsoft/STL/blob/main/stl/inc/algorithm#L8060) and/or MSVC resolves the `begin()` and `end()` unexpectedly to the more restricted *const* case? A quick google search seems to agree with this ([here](https://stackoverflow.com/questions/48250560/msvc-error-c2593-when-overloading-const-and-non-const-conversion-operator-return) and [here](https://stackoverflow.com/questions/55911858/msvc-inference-of-pointer-to-member-function-with-const-and-non-const-overloads)). This just seems very odd, considering one would expect Microsoft has some unit tests to catch this exact thing.

Reply:
I didn't realize the compile error was from the GitHub Actions API!

I've tried to reproduce the compile error on Godbolt running MSVC 19.14, but unfortunately I can't reproduce it - it compiles without error. I wonder if there's some way I can easily run MSVC 19.34 myself to try out my reproduction case: https://godbolt.org/z/ocMb3Gcdh

Reply:
Updating to the latest version of said GitHub action also updates the default to be *MSVC 19.34.31942.0*.

Reply:
I tried to click on your link. Here it by default had *x64 msvc v19.14 (WINE)*. When I switched it to *x64 msvc v19 latest*. As far as I can tell, the *WINE* part means, that it is Microsoft's compiler but run on a linux machine rather than a Windows machine?

Reply:
With a binary search on the provided *godbolt* link, the issue was introduced with version 19.28 of MSVC x64. There are multiple version numbers to think about. Looking at the Wikipedia page, this change was accompanied with:

|                                    | Working       | Failing           |
|-----------------------------|-------------------|-------------------|
| Product Version      | 16.7               | 16.8.1            |
| Tool chain number | 14.27             | 14.28             |
| MSC version            | 19.27.29112 | 19.28.29333 |

Reply:
On second thought, my hack does actually make some sense. What should the *const* in `const iter_base` refer to? Should it mean I cannot change where it points to, the underlying value, or both? In other words, is a `const iter_base` equivalent to `T* const`, `const T*` or `const T* const*`?